Deprecate returning non-string values from a user output handler

DanielEScherzer opened the pull request:
https://wiki.php.net/rfc/deprecations_php_8_4

TimWolla commented on main/output.c:
```diff
@@ -959,17 +960,33 @@ static inline php_output_handler_status_t php_output_handler_op(php_output_handl
+						"Returning a non-string result from user output handler %s is deprecated",
+						ZSTR_VAL(handler->name)
+					);
+					OG(flags) = old_flags;
```
Do you also need to check `EG(exception)` for a throwing error handler?

DanielEScherzer replied:
Actually looks like we don't actually need to check EG(exception) because I switched to only disabling the specific handler that was problematic

Girgias reviewed (commented):
Thank you for tackling this! I added a few comments and I think the reason you don't get the "correct" name for anonymous functions should be fixed with https://github.com/php/php-src/pull/19011

DanielEScherzer commented on main/output.c:
```diff
@@ -959,17 +961,48 @@ static inline php_output_handler_status_t php_output_handler_op(php_output_handl
+					/* user handler may have returned TRUE */
+					status = PHP_OUTPUT_HANDLER_NO_DATA;
+					if (Z_TYPE(retval) != IS_FALSE && Z_TYPE(retval) != IS_TRUE) {
+						convert_to_string(&retval);
```
Done, also tested for a stringable object

DanielEScherzer commented:
> Thank you for tackling this! I added a few comments and I think the reason you don't get the "correct" name for anonymous functions should be fixed with #19011

Rebased after that patch was merged and indeed the anonymous functions no longer just say "Closure::__invoke", https://github.com/php/php-src/pull/18932/commits/87756d772c08d43b414c31f2a4f6b287fa8e55eb

Girgias commented on Zend/tests/declare/gh18033_2.phpt:
```diff
@@ -13,4 +13,5 @@ ob_start(function() {
     );
```
This test should be fixed so that we don't remove it after the deprecation.
```suggestion
    );
    return '';
```

DanielEScherzer replied:
Yeah, I didn't do that at first because these tests were good for making sure I didn't have any errors in my handling of non-strings in a variety of cases, done

Girgias commented on Zend/tests/concat/bug79836_2.phpt:
```diff
@@ -21,5 +21,8 @@ $x = $c . $xxx;
 echo $x . "\n";
 ?>
---EXPECT--
+--EXPECTF--
+Deprecated: X::__toString(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
```
Ditto change the output handler to return an empty string

Girgias commented on Zend/tests/concat/bug79836_1.phpt:
```diff
@@ -14,5 +14,8 @@ $x = $c . $x;
 echo "Done\n";
 ?>
---EXPECT--
+--EXPECTF--
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
+Deprecated: ob_end_clean(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
```
Ditto

Girgias commented on Zend/tests/concat/bug79836.phpt:
```diff
@@ -14,5 +14,10 @@ $c .= [];
 echo $counter . "\n";
 ?>
---EXPECT--
+--EXPECTF--
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
+Deprecated: ob_end_clean(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
```
Ditto

Girgias commented on Zend/tests/gh11189.phpt:
```diff
@@ -29,4 +29,6 @@ while (1) {
 ?>
 --EXPECTF--
 Success
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
```
Ditto

Girgias commented on Zend/tests/gh11189_1.phpt:
```diff
@@ -29,4 +29,6 @@ while (1) {
 ?>
 --EXPECTF--
 Success
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
```
Ditto

Girgias commented on Zend/tests/gh16408.phpt:
```diff
@@ -12,5 +12,10 @@ $c .= [];
 echo $counter . "\n";
 ?>
---EXPECT--
+--EXPECTF--
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
+
+Deprecated: ob_end_clean(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
 3
```
Ditto

Girgias commented on ext/session/tests/user_session_module/bug61728.phpt:
```diff
@@ -40,5 +40,6 @@ class MySessionHandler implements SessionHandlerInterface {
 ?>
---EXPECT--
+--EXPECTF--
+Deprecated: ob_end_flush(): Returning a non-string result from user output handler output_html is deprecated in %s on line %d
```
Ditto

Girgias commented on main/output.c:
```diff
@@ -1225,6 +1265,19 @@ static int php_output_stack_pop(int flags)
+		int handler_count = php_output_get_level();
+		if (handler_count) {
+			php_output_handler **handlers = (php_output_handler **) zend_stack_base(&OG(handlers));
+			for (int iii = 0; iii < handler_count; ++iii) {
```
Please use something more descriptive than `iii` for the variable name

Girgias commented on sapi/cli/tests/gh8827-002.phpt:
```diff
@@ -32,10 +32,20 @@ print "STDOUT:\n";
---EXPECT--
+--EXPECTF--
 STDIN:
+
+Deprecated: main(): Returning a non-string result from user output handler {closure:%s:%d} is deprecated in %s on line %d
```
Ditto fix output handler

Girgias reviewed (changes requested):
Few nits, but looks mostly good :)

bukka commented:
Or lets leave it as you put already some effort into this and we will see how it goes. It was strange edge case anyway and might be a bit unusual for people to do that.
